File: lib/api/heatmap.ts
import { apiClient } from './client';

export interface HeatmapConfig {
  id: string;
  user_id: string;
  watch_list_id: string;
  name: string;
  size_metric: 'market_cap' | 'volume' | 'avg_volume' | 'reddit_mentions' | 'reddit_popularity';
  color_metric: 'price_change_pct' | 'volume_change_pct' | 'reddit_rank' | 'reddit_trend';
  time_period: '1D' | '1W' | '1M' | '3M' | '6M' | 'YTD' | '1Y' | '5Y';
  color_scheme: 'red_green' | 'heatmap' | 'blue_red' | 'custom';
  label_display: 'symbol' | 'symbol_change' | 'full';
  layout_type: 'treemap' | 'grid';
  filters: Record<string, any>;
  color_gradient?: Record<string, string>;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface HeatmapTile {
  symbol: string;
  name: string;
  asset_type: string;
  size_value: number;
  size_label: string;
  color_value: number;
  color_label: string;
  current_price: number;
  price_change: number;
  price_change_pct: number;
  volume?: number;
  market_cap?: number;
  prev_close?: number;
  exchange: string;
  // Reddit data
  reddit_rank?: number;
  reddit_mentions?: number;
  reddit_popularity?: number;
  reddit_trend?: 'rising' | 'falling' | 'stable';
  reddit_rank_change?: number;
  // User watchlist data
  notes?: string;
  tags: string[];
  target_buy_price?: number;
  target_sell_price?: number;
}

export interface HeatmapData {
  watch_list_id: string;
  watch_list_name: string;
  config_id?: string;
  config_name?: string;
  size_metric: string;
  color_metric: string;
  time_period: string;
  color_scheme: string;
  tiles: HeatmapTile[];
  tile_count: number;
  min_color_value: number;
  max_color_value: number;
  generated_at: string;
}

export const heatmapAPI = {
  // Get heatmap data
  async getHeatmapData(
    watchListId: string,
    configId?: string,
    overrides?: {
      size_metric?: string;
      color_metric?: string;
      time_period?: string;
    }
  ): Promise<HeatmapData> {
    const params = new URLSearchParams();
    if (configId) params.append('config_id', configId);
    if (overrides?.size_metric) params.append('size_metric', overrides.size_metric);
    if (overrides?.color_metric) params.append('color_metric', overrides.color_metric);
    if (overrides?.time_period) params.append('time_period', overrides.time_period);

    const queryString = params.toString();
    const url = `/watchlists/${watchListId}/heatmap${queryString ? `?${queryString}` : ''}`;
    return apiClient.get(url);
  },

  // Get all configs for a watch list
  async getConfigs(watchListId: string): Promise<{ configs: HeatmapConfig[] }> {
    return apiClient.get(`/watchlists/${watchListId}/heatmap/configs`);
  },

  // Create new config
  async createConfig(
    watchListId: string,
    config: {
      name: string;
      size_metric: string;
      color_metric: string;
      time_period: string;
      color_scheme?: string;
      label_display?: string;
      layout_type?: string;
      filters?: Record<string, any>;
      color_gradient?: Record<string, string>;
      is_default?: boolean;
    }
  ): Promise<HeatmapConfig> {
    return apiClient.post(`/watchlists/${watchListId}/heatmap/configs`, {
      watch_list_id: watchListId,
      ...config,
    });
  },

  // Update config
  async updateConfig(
    watchListId: string,
    configId: string,
    config: Partial<HeatmapConfig>
  ): Promise<HeatmapConfig> {
    return apiClient.put(`/watchlists/${watchListId}/heatmap/configs/${configId}`, config);
  },

  // Delete config
  async deleteConfig(watchListId: string, configId: string): Promise<void> {
    return apiClient.delete(`/watchlists/${watchListId}/heatmap/configs/${configId}`);
  },
};

File: backend/models/heatmap.go
package models

import (
	"time"
)

// HeatmapConfig represents a saved heatmap configuration
type HeatmapConfig struct {
	ID                string                 `json:"id" db:"id"`
	UserID            string                 `json:"user_id" db:"user_id"`
	WatchListID       string                 `json:"watch_list_id" db:"watch_list_id"`
	Name              string                 `json:"name" db:"name"`
	SizeMetric        string                 `json:"size_metric" db:"size_metric"`
	ColorMetric       string                 `json:"color_metric" db:"color_metric"`
	TimePeriod        string                 `json:"time_period" db:"time_period"`
	ColorScheme       string                 `json:"color_scheme" db:"color_scheme"`
	LabelDisplay      string                 `json:"label_display" db:"label_display"`
	LayoutType        string                 `json:"layout_type" db:"layout_type"`
	FiltersJSON       map[string]interface{} `json:"filters" db:"filters_json"`
	ColorGradientJSON map[string]string      `json:"color_gradient,omitempty" db:"color_gradient_json"`
	IsDefault         bool                   `json:"is_default" db:"is_default"`
	CreatedAt         time.Time              `json:"created_at" db:"created_at"`
	UpdatedAt         time.Time              `json:"updated_at" db:"updated_at"`
}

// HeatmapTile represents a single tile in the heatmap
type HeatmapTile struct {
	Symbol    string `json:"symbol"`
	Name      string `json:"name"`
	AssetType string `json:"asset_type"`
	Sector    string `json:"sector"`

	// Size value (what determines tile size)
	SizeValue float64 `json:"size_value"`
	SizeLabel string  `json:"size_label"` // e.g., "$1.2B", "15.3M shares"

	// Color value (what determines tile color)
	ColorValue float64 `json:"color_value"`
	ColorLabel string  `json:"color_label"` // e.g., "+5.2%", "-2.1%"

	// Current price info
	CurrentPrice   float64 `json:"current_price"`
	PriceChange    float64 `json:"price_change"`
	PriceChangePct float64 `json:"price_change_pct"`

	// Additional metadata for tooltip
	Volume    *int64   `json:"volume,omitempty"`
	MarketCap *float64 `json:"market_cap,omitempty"`
	PrevClose *float64 `json:"prev_close,omitempty"`
	Exchange  string   `json:"exchange"`

	// Reddit data (from reddit_heatmap_daily table)
	RedditRank       *int     `json:"reddit_rank,omitempty"`        // Current Reddit rank (1 = #1 trending)
	RedditMentions   *int     `json:"reddit_mentions,omitempty"`    // Total mentions
	RedditPopularity *float64 `json:"reddit_popularity,omitempty"`  // Popularity score (0-100)
	RedditTrend      *string  `json:"reddit_trend,omitempty"`       // "rising", "falling", "stable"
	RedditRankChange *int     `json:"reddit_rank_change,omitempty"` // Rank change vs 24h ago

	// User's custom data from watch list
	Notes           *string  `json:"notes,omitempty"`
	Tags            []string `json:"tags,omitempty"`
	TargetBuyPrice  *float64 `json:"target_buy_price,omitempty"`
	TargetSellPrice *float64 `json:"target_sell_price,omitempty"`
}

// HeatmapData represents the complete heatmap data
type HeatmapData struct {
	WatchListID   string `json:"watch_list_id"`
	WatchListName string `json:"watch_list_name"`
	ConfigID      string `json:"config_id,omitempty"`
	ConfigName    string `json:"config_name,omitempty"`

	// Configuration used
	SizeMetric  string `json:"size_metric"`
	ColorMetric string `json:"color_metric"`
	TimePeriod  string `json:"time_period"`
	ColorScheme string `json:"color_scheme"`

	// Tiles data
	Tiles     []HeatmapTile `json:"tiles"`
	TileCount int           `json:"tile_count"`

	// Metadata for color scale
	MinColorValue float64 `json:"min_color_value"`
	MaxColorValue float64 `json:"max_color_value"`

	// Timestamp
	GeneratedAt time.Time `json:"generated_at"`
}

// Request/Response DTOs

// CreateHeatmapConfigRequest for saving a new config
type CreateHeatmapConfigRequest struct {
	WatchListID   string                 `json:"watch_list_id" binding:"required"`
	Name          string                 `json:"name" binding:"required,min=1,max=255"`
	SizeMetric    string                 `json:"size_metric" binding:"required,oneof=market_cap volume avg_volume reddit_mentions reddit_popularity"`
	ColorMetric   string                 `json:"color_metric" binding:"required,oneof=price_change_pct volume_change_pct reddit_rank reddit_trend"`
	TimePeriod    string                 `json:"time_period" binding:"required,oneof=1D 1W 1M 3M 6M YTD 1Y 5Y"`
	ColorScheme   string                 `json:"color_scheme" binding:"oneof=red_green heatmap blue_red custom"`
	LabelDisplay  string                 `json:"label_display" binding:"oneof=symbol symbol_change full"`
	LayoutType    string                 `json:"layout_type" binding:"oneof=treemap grid"`
	Filters       map[string]interface{} `json:"filters"`
	ColorGradient map[string]string      `json:"color_gradient,omitempty"`
	IsDefault     bool                   `json:"is_default"`
}

// UpdateHeatmapConfigRequest for updating existing config
type UpdateHeatmapConfigRequest struct {
	Name          string                 `json:"name" binding:"min=1,max=255"`
	SizeMetric    string                 `json:"size_metric" binding:"oneof=market_cap volume avg_volume reddit_mentions reddit_popularity"`
	ColorMetric   string                 `json:"color_metric" binding:"oneof=price_change_pct volume_change_pct reddit_rank reddit_trend"`
	TimePeriod    string                 `json:"time_period" binding:"oneof=1D 1W 1M 3M 6M YTD 1Y 5Y"`
	ColorScheme   string                 `json:"color_scheme" binding:"oneof=red_green heatmap blue_red custom"`
	LabelDisplay  string                 `json:"label_display" binding:"oneof=symbol symbol_change full"`
	LayoutType    string                 `json:"layout_type" binding:"oneof=treemap grid"`
	Filters       map[string]interface{} `json:"filters"`
	ColorGradient map[string]string      `json:"color_gradient,omitempty"`
	IsDefault     bool                   `json:"is_default"`
}

// GetHeatmapDataRequest for generating heatmap data
type GetHeatmapDataRequest struct {
	WatchListID string `json:"watch_list_id"`
	ConfigID    string `json:"config_id,omitempty"` // If empty, use default config

	// Override config settings (optional)
	SizeMetric  string                 `json:"size_metric,omitempty"`
	ColorMetric string                 `json:"color_metric,omitempty"`
	TimePeriod  string                 `json:"time_period,omitempty"`
	Filters     map[string]interface{} `json:"filters,omitempty"`
}
